packages/labs/src/frameworks/react/create-component.ts

```diff
@@ -0,0 +1,216 @@
+
+type Constructor<T> = {new (): T};
+
+/**
+ *  Creates a React component from a CustomElement.
+ */
+export const createComponent = <I extends HTMLElement, E>(
+  React: typeof ReactModule,
+  tagName: string,
+  elementClass: Constructor<I>,
+  events?: E
+) => {
+  const Component = React.Component;
+  const createElement = React.createElement;
+
+  // Props the user is allowed to use, includes standard attributes, children,
+  // ref, as well as special event and element properties.
+  type UserProps = React.PropsWithChildren<
+    React.PropsWithRef<Partial<I> & Events<E>>
+  >;
+
+  // Props used by this component wrapper. This is the UserProps and the
+  // special `__forwardedRef` property. Note, this ref is special because
+  // it's both needed in this component to get access to the rendered element
+  // and must fulfill any ref passed by the user.
+  type ComponentProps = UserProps & {
+    __forwardedRef?: React.Ref<unknown>;
+  };
+
+  // List of properties/events which should be specially handled by the wrapper
+  // and not handled directly by React.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const elementPropsMap: {[index: string]: any} = {...(events ?? {})};
+  for (const p in elementClass.prototype) {
+    if (!(p in HTMLElement.prototype)) {
+      elementPropsMap[p] = true;
+    }
+  }
+
+  class ReactComponent extends Component<ComponentProps> {
+    private _element: I | null = null;
+
+    private _elementProps!: typeof elementPropsMap;
+
+    private _ref = (element: I | null) => {
+      this._element = element;
+      // Fulfill any ref we receive in our props.
+      // The `ref` prop is special and cannot be accessed without an error
+      // so use `__forwardedRef`.
+      const forwardedRef = this.props?.__forwardedRef;
+      if (forwardedRef) {
+        // Note, there are 2 kinds of refs and we manually fulfill them here.
+        // There is no built in React API for this.
+        if (typeof forwardedRef === 'function') {
+          (forwardedRef as (e: I | null) => void)(element);
+        } else {
+          // This has to be `any` because `current` is readonly.
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          (forwardedRef as any).current = element;
+        }
+      }
+    };
+
+    private _updateElement(oldProps?: ComponentProps) {
+      if (this._element === null) {
+        return;
+      }
+      // Set element properties to the values in `this.props`
+      for (const prop in this._elementProps) {
+        setProperty(
+          this._element,
+          prop,
+          this.props[prop as keyof ComponentProps],
+          oldProps ? oldProps[prop as keyof ComponentProps] : undefined,
+          events
+        );
+      }
+      // Note, the spirit of React might be to "unset" any old values that
+      // are no longer included; however, there's no reasonable value to set
+      // them to so we just leave the previous state as is.
```

Comment:
I don't think so. I believe users write LitElement's with `@property({type: String}) foo = ''` and expect that value to never be `undefined`. They shouldn't but I think this is not uncommon.

Reply:
This behavior needs to be documented then. I'm still not sure setting `undefined` isn't the right behavior though. If a user of an elements sets and then "unsets" a property, isn't that user at least expecting something to happen on unset?

And to be clear, this can only happen with spread, right? If a prop is explicitly set to undefined it shows up in the props object?

Reply:
This is my remaining concern on this PR. Make a note of it in the docs on `createComponent`? I think it's only an issue with spread.